### ISSUE TITLE
Allowing type guards to perform incomplete check

### ISSUE THREAD

**Issue by yw662** [microsoft/TypeScript#49767]
# Suggestion

## 🔍 Search Terms

type guard, narrowing

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Usually, if a type guard (isFoo, let's say) returns false, the value will be narrowed to `Exclude<T, foo>`, requiring a type guard must always return true if the value if of the type.

However, if we are in the situation that the guard only detect a subset of the type -- it is still a type guard since it can narrow the value `foo` as `Foo`, and `isFoo(foo): foo is Foo` is still semantically correct (in English), it is indeed `MoreSpecificFoo` though, but a `MoreSpecificFoo` is a `Foo` -- wrongly marking it as a type guard may resulting in troubles:

```ts
const isFoo = (v: number | string): v is number => v === 1

const f = (v: number | string) => {
    if(isFoo(v)) return v
    return v.split('')
}

f(2)
```

The suggestion is to allow this kind of "half guards", where `true` means `true` and `false` means `unknown`, or vice versa, like:
```ts
const isFoo = (v: number | string): "v is number on true, but no assertion on false" => v === 1

const f = (v: number | string) => {
    if(isFoo(v)) return v
    return v.split('')
}

f(2)
```

## 📃 Motivating Example

Those "half guards" can be useful on positive narrowing, and like `isFoo: v is 1` can solve the above problem, but it is not always possible to do so.

## 💻 Use Cases

Just as described above.

**Comment by fatcerberus** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/15048
